add proptable #19

nickeubank opened the pull request:
still needs tests

nalimilan commented on src/freqtable.jl:
```diff
@@ -227,3 +227,59 @@ end
+
+"""
+    proptable(x::AbstractVector...; 
+              margin::Union{Nothing, Integer, Vector{Integer}} = nothing,
```
I wonder whether we should call this `dims` for consistency with other reductions. Then we should also use a keyword argument for `prop`. @bkamins?

bkamins replied:
A funny thing is that when we introduced `prop` we have discussed all this I think. The decision was to go with `prop` over `freqtable`, but I am OK to add `proptable`. Also regarding making `margin` a kwarg I think it is a good direction also for `prop` as it is consistent with the same type of change in Base.

Now regarding `dims` vs `margin` it was a conscious decision as `margin` is a complement of `dims`, but maybe we can change this (back then we decided that `margin` was a more natural way to specify along which dimension you want sums to be equal to `1`). Also note that if we do it we should probably accept an iterable not a `Vector{Integer}` only (e.g. now a tuple is excluded by the signature)

nickeubank replied:
I was just trying to ensure reverse compatibility. Happy to go any direction the team wants with this.

nickeubank replied:
@bkamins what's best practice for allowing any iterable? Just drop the type-check? There is not abstract type for iterables is there?

bkamins replied:
So my opinion is:
* add `proptable`;
* make a kwarg with the same mechanics as in Base (e.g. accepting a tuple)
* the `margin`/`dims` decision is more delicate; `dims` would say which dimensions we want to aggregate, `margin` is the opposite - is says in which dimension we want the sum equal to `1`; not sure which is more intuitive; earlier we chose `margin` from these two because R uses `margin`; so - I would slightly favor `margin`, but this is not a strong opinion

bkamins replied:
> what's best practice for allowing any iterable?

Base does not do a type restriction in the signature, but simply tries to iterate `dims`, see `reduced_indices` function in reducedim.jl in Base to check how it is handled (the mechanics there is more complex and probably we do not need something as advanced here). The `dims` kwarg is called `region` there.

nickeubank replied:
As someone who does more statistics than linear algebra, `margin` (e.g. marginal distributions) makes more sense in my head than `dims`. 

> make a kwarg with the same mechanics as in Base (e.g. accepting a tuple)

Do you mean the general idea of requiring margins to be either a single `Integer` or a `Tuple{Integer}`, or is there an actual implementation of `prop` you want to mirror? I'm not finding a `prop` in Base or in Statistics.

bkamins replied:
I mean:

>  the general idea of requiring margins to be either a single `Integer` or a `Tuple{Integer}`

and then `prop` should be redesigned to follow this rule.

However note that we should also accept any iterable of `Integer` (not only tuple or vector) as this is what Base does. Is this clear now?

nalimilan replied:
OK, let's keep `margin` then. With keyword arguments, we could even support both `margin` and `dims` if we wanted.

bkamins replied:
having both `margin` and `dims` as possible options is a very nice idea. Of course they should be exclusive.

bkamins replied:
@nickeubank Have you considered adding `dims` also?

nickeubank replied:
how exactly do dims work on, say, a 3 dimensional array?

bkamins replied:
They are exactly a complement of `margin`. I.e. if you have 3 dimensions then `margin=(1,2)` is the same as `dims=3`.
Or with a matrix, omitting `margin` is the same as `dims=(1,2)`.

`margin` says in which dimensions we ensure that the sum is `1`. `dims` says over which dimensions we aggregate. So if you have rows summing to 1 (`margin=1`) this means that you aggregate over columns (`dims=2`)

nalimilan commented on src/freqtable.jl:
```diff
@@ -227,3 +227,59 @@ end
+            
+function proptable(d, x::Symbol...; kwargs...)
+
+    if (:margin in keys(kwargs))                 
```
Why not have `margins=nothing` in the signature and check whether it's `nothing`, as above?

nickeubank replied:
@nalimilan Oh! right of course.

nickeubank commented:
Yup! Will do this week

On Sat, Mar 23, 2019 at 9:53 AM Bogumił Kamiński <notifications@github.com>
wrote:

> *@bkamins* commented on this pull request.
> ------------------------------
>
> In src/freqtable.jl
> <https://github.com/nalimilan/FreqTables.jl/pull/38#discussion_r268397455>
> :
>
> > @@ -227,3 +227,59 @@ end
>
>  prop(tbl::NamedArray{<:Number}, margin::Integer...) =
>      NamedArray(prop(convert(Array, tbl), margin...), tbl.dicts, tbl.dimnames)
> +
> +
> +
> +"""
> +    proptable(x::AbstractVector...;
> +              margin::Union{Nothing, Integer, Vector{Integer}} = nothing,
>
> I mean:
>
> the general idea of requiring margins to be either a single Integer or a
> Tuple{Integer}
>
> and then prop should be redesigned to follow this rule.
>
> However note that we should also accept any iterable of Integer (not only
> tuple or vector) as this is what Base does. Is this clear now?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nalimilan/FreqTables.jl/pull/38#discussion_r268397455>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJPC7YMS2lXGip1yQfpI9IxMRRdeSE7Oks5vZj_jgaJpZM4cDuM9>
> .
>

nickeubank commented:
Think addresses everything, except that tests require Iris.csv, which I don't think is available any more (https://github.com/JuliaData/DataFrames.jl/issues/1760)

bkamins commented:
Use `joinpath(dirname(pathof(DataFrames)), "../docs/src/assets/iris.csv")` to get iris.csv file.

bkamins commented on src/freqtable.jl:
```diff
@@ -154,14 +154,17 @@ function freqtable(d, x::Symbol...; args...)
 
 """
-    prop(tbl::AbstractArray{<:Number}, [margin::Integer...])
+    prop(tbl::AbstractArray{<:Number}, [margin::Union{Integer, Tuple{Integer}}...])
```
Below you write we accept any iterable. Also I think `...` is not needed any more. Base docs would simply write `margin` without type qualification.

bkamins replied:
We also allow not to pass `margin`

bkamins commented on src/freqtable.jl:
```diff
@@ -219,11 +222,58 @@ julia> sum(pt, (1, 2))
 prop(tbl::AbstractArray{<:Number}) = tbl / sum(tbl)
 
-function prop(tbl::AbstractArray{<:Number,N}, margin::Integer...) where N
+function prop(tbl::AbstractArray{<:Number,N}, margin) where N
```
@nalimilan do we want to support deprecation for the old syntax? (technically we should)

nalimilan replied:
Yes, certainly. But we agreed to make `margin` a keyword argument, right?

Also, I realize it should be `margins` (plural) for consistency with `dims`.

bkamins replied:
All agreed (and this is what I had in mind, except plural form, because I did not think about it - but I agree it should be plural)

bkamins commented on src/freqtable.jl:
```diff
@@ -219,11 +222,58 @@ julia> sum(pt, (1, 2))
-    NamedArray(prop(convert(Array, tbl), margin...), tbl.dicts, tbl.dimnames)
+prop(tbl::NamedArray{<:Number}, margin) =
+    NamedArray(prop(convert(Array, tbl), margin), tbl.dicts, tbl.dimnames)
+
```
Extra empty lines are not needed. One is enough.

bkamins commented on src/freqtable.jl:
```diff
@@ -219,11 +222,58 @@ julia> sum(pt, (1, 2))
+              weights::AbstractVector{<:Real} = UnitWeights(),
+              subset::Union{Nothing, AbstractVector{Int}, AbstractVector{Bool}} = nothing])
+        
+    proptable(d, x::Symbol...; margin = nothing,
```
what is the contract for `d` here? Also I would add a newline after `;` to match the style above.

nickeubank commented:
More tests soon...

Also need to add docs to freqtable. Could be separate issue if anyone wants but seems like need to be done.

nickeubank commented:
@bkamins still getting build error from iris file even with tweak:

> ERROR: LoadError: LoadError: ArgumentError: "/home/travis/.julia/packages/DataFrames/IKMvt/src/../docs/src/assets/iris.csv" is not a valid file

Though I think path is what you suggested: https://github.com/nalimilan/FreqTables.jl/blob/b15d626342fee60a0bbb640579fa2058c5ceee8c/test/freqtable.jl#L147

bkamins commented:
If it is OK with you can you please update everything in one shot?
If you want you can split `dims` kwarg to a separate PR and submit it after this is merged.

nickeubank commented:
yup will do, though prob not for little while. baby woke up. :)

bkamins commented:
Are you testing against DataFrames.jl master? The file was added after the last release.

nickeubank commented:
ah, ok. I think travis is building against last release.

bkamins commented:
OK - we can probably wait for DataFrames.jl release (hopefully we can get it in a few weeks - we are almost there; @nalimilan what do you think?)

nalimilan commented on src/freqtable.jl:
```diff
@@ -219,11 +221,60 @@ julia> sum(pt, (1, 2))
+              weights::AbstractVector{<:Real} = UnitWeights(),
+              subset::Union{Nothing, AbstractVector{Int}, AbstractVector{Bool}} = nothing])
+
+Create frequency table of proportions from vectors or table columns with margins generated for dimensions specified by `margin`. If `margin` is omitted proportions over the whole table 
```
Please cut lines at 92 chars. Also "a frequency table".

nalimilan commented on src/freqtable.jl:
```diff
@@ -219,11 +221,60 @@ julia> sum(pt, (1, 2))
+In particular when `margin` is `1` row proportions,
+and when `margin` is `2` column proportions are calculated.
+
+`d` must be a Table.        
```
Better be precise, especially since "table" can be confused with "frequency table":
```
`d` can be any type of table supported by the [Tables.jl](https://github.com/JuliaData/Tables.jl) interface.
```
I'd also rename `d` to `t` since it's a table, and `x` to `cols` or `vars`.

nalimilan commented on src/freqtable.jl:
```diff
@@ -219,11 +221,60 @@ julia> sum(pt, (1, 2))
+        
+`margin` must be an Integer or an iterable of Integers.        
+        
+Calculating `sum` over the result of `prop` over dimensions that are complement of `margin`
```
This doesn't sound very clear to me. Maybe this? And move it after the example about `margin=1` and `margin=2`.
```
More generally, the resulting array will have sums equal to one
for all dimensions not specified in `margin`.
```

nalimilan commented on src/freqtable.jl:
```diff
@@ -219,11 +221,60 @@ julia> sum(pt, (1, 2))
+end
+            
+function proptable(d, x::Symbol...; margin=nothing, kwargs...)
+    Tables.istable(d) || throw(ArgumentError("data must be a table"))
```
`freqtable` already does this check, right?

nalimilan commented on test/freqtable.jl:
```diff
@@ -190,3 +193,23 @@ intft = freqtable(df, :A, :B)
 @test_throws BoundsError intft[101,"x"]
 @test intft[Name(101),"x"] == 1
+    
+# Proptable
```
```suggestion
# proptable
```

nalimilan commented:
> OK - we can probably wait for DataFrames.jl release (hopefully we can get it in a few weeks - we are almost there; @nalimilan what do you think?)

Can we use RDatasets instead? That would be more robust than relying on a semi-private file in DataFrames.jl.

bkamins commented:
Up to you as it is your package 😄, but RDatasets.jl is a heavy dependency and e.g. I do not have it installed at all for this reason (so maybe it is simplest to have the data set in this package - not sure).

nalimilan commented:
Note that it's just for testing. We could also copy the data in the package, but if we do that probably better have it inline so that we don't require CSV.jl either (and maybe use a smaller data set then).

bkamins commented:
> and maybe use a smaller data set then

This is what I would actually prefer (minimizing dependencies simplifies user's life).

nickeubank commented:
OK, saving `dims` for another day, but:

- added freqtable docs and examples
- re-wrote tests to not use Iris csv, removed CSV from requirements
- changed from `margin` to `margins`

nickeubank commented:
Side note: I have no idea how doctests works, so... just tried to mimic pre-existing examples.

nickeubank commented:
squash and rebase

bkamins commented on src/freqtable.jl:
```diff
@@ -219,11 +265,60 @@ julia> sum(pt, (1, 2))
-    lo, hi = extrema(margin)
-    (lo < 1 || hi > N) && throw(ArgumentError("margin must be a valid dimension"))
-    tbl ./ sum(tbl, dims=tuple(setdiff(1:N, margin)...)::NTuple{N-length(margin),Int})
+function prop(tbl::AbstractArray{<:Number,N}, margins) where N
```
`margins` should be a keyword argument to match `proptable`.

nalimilan commented on src/freqtable.jl:
```diff
@@ -81,6 +81,50 @@ function _freqtable(x::Tuple,
+4     │ 1
+
+julia> df = DataFrame(x=[1, 2, 2, 2], y=[1, 2, 1, 2]);
+julia> freqtable(df, :x, :y)
```
Missing empty line (will probably make doctests fail).

nalimilan commented on test/freqtable.jl:
```diff
@@ -190,3 +196,23 @@ intft = freqtable(df, :A, :B)
+df = DataFrame(x = [1, 2, 1, 2], y = [1, 1, 2, 2], z = ["a", "a", "c", "d"])
+
+tab = proptable(df, :x, :z)    
+@test tab == [0.25 0.25 0.0
```
Can you also test names? If they are incorrect the whole point of frequency tables collapses... :-)

nalimilan commented on test/freqtable.jl:
```diff
@@ -190,3 +196,23 @@ intft = freqtable(df, :A, :B)
+@test_throws ArgumentError proptable(df)
+@test_throws ArgumentError proptable(nothing, :x, :y)
+@test_throws MethodError proptable(df, :x, :y, 1, 2)
+
```
No need to empty lines.

nalimilan commented on src/freqtable.jl:
```diff
@@ -308,17 +313,9 @@ function proptable(x::AbstractVector...;
                    skipmissing::Bool = false,
                    weights::AbstractVector{<:Real} = UnitWeights(),
                    subset::Union{Nothing, AbstractVector{Int}, AbstractVector{Bool}} = nothing)
```
Use short function form (with `=`) for one-liners. Same below.

nalimilan commented on src/freqtable.jl:
```diff
@@ -145,26 +190,29 @@ freqtable(x::AbstractCategoricalVector...; skipmissing::Bool = false,
-dimensions specified by `margin`.
-If `margin` is omitted proportions over the whole `tbl` are computed.
+dimensions specified by `margins`.
+If `margins` is `nothin`, proportions over the whole `tbl` are computed.
```
```suggestion
If `margins` is `nothing` (the default), proportions over the whole `tbl` are computed.
```

nalimilan commented on src/freqtable.jl:
```diff
@@ -145,26 +190,29 @@ freqtable(x::AbstractCategoricalVector...; skipmissing::Bool = false,
+In particular when `margins` is `1` row proportions,
+and when `margins` is `2` column proportions are calculated.
+        
+`margins` must be `nothing`, an `Integer`, or an iterable of `Integer`s.        
```
Put this before explanations about the effect of these values. Same in the other docstring.

nalimilan commented on src/freqtable.jl:
```diff
@@ -227,3 +227,59 @@ end
+Calculating `sum` over the result of `prop` over dimensions that are complement of `margin`
+produces `AbstractArray` containing only `1.0`, see last example below.
+
+
```
This one still applies.

nickeubank replied:
sry, lost track of this. Will add this week

nickeubank commented:
OK, think I finally got it pinned down. 

One note: I don't think the doctests are being run by travis -- I found an outstanding bug in the `prop` doctests when I ran them by hand (`dims` is keyword-only now for `sum`, the doctest was using as an unnamed argument).

nalimilan commented:
Thanks. I've edited the docstrings trying to make them clearer, let me know what you think. In particular, I find that saying that proportions sum to one over `margins` explains quite well what the function does. That's the opposite to what `dims` does (it indicates dimensions to keep), but I don't think that's a problem as long as we show the example using `sum(..., dims=dims)`.

nickeubank commented:
That seems great, thanks! 

The one other way it occurs to me we could summarize it is by pointing out that `margin` is short for marginal distributions, and that `margins=1` is just giving the marginal distributions for each row, and `margins=2` gives marginal distributions for each column.

nalimilan commented:
Thanks.

Actually, the doctests aren't run because there's no manual.
